### PR TITLE
`ClientRequestContext#cancel` cancels the associated request immediately

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

In order to handle https://github.com/line/armeria/issues/4591, I propose that we first define an API which allows users to cancel a request. Currently,  `ClientRequestContext#cancel` is invoked once `CancellationScheduler#start` is invoked. I propose to change the behavior of `ClientRequestContext#cancel`  such that the associated request is aborted immediately. Once this API is available, it would be trivial to implement `ResponseTimeoutMode` by adjusting where to call `CancellationScheduler#start`. Additionally, users may easily implement their own timeout mechanism if they would like.

Modifications:

- `CancellationScheduler` related changes
  - `DefaultCancellationScheduler` is refactored to be lock based instead of event loop based. The reasoning for this change was for the scenario where the request execution didn't reach the event loop yet. i.e. If a user calls `ctx.cancel` in the decorator, a connection shouldn't be opened.
    - e.g. https://github.com/line/armeria/blob/59aa40a59e1f1122716e70f9f1d6f1402a6aae0e/core/src/test/java/com/linecorp/armeria/client/ContextCancellationTest.java#L90-L116
  - `CancellationScheduler#updateTask` is introduced. This API updates the cancellation task if the scheduler isn't invoked yet. If the scheduler is invoked already, the cancellation task will be executed eventually. This API allows `ctx.cancel` to attempt cancellation depending on which stage the request is at. For instance, at the decorators only `req.abort` needs to be called but at the request write stage, the cancellation task may need to send a reset signal.
  - Misc. an infinite timeout is internally represented as `Long.MAX_VALUE` instead of `0`
- `AbstractHttpRequestHandler` related changes
  - `CancellationTask` in `AbstractHttpRequestHandler`, `HttpResponseWrapper`, `AbstractHttpResponseHandler` is modified to be scheduled inside an event loop. The reasoning is that `ctx.cancel`, and hence `CancellationTask#run` can be invoked from any thread.
  - There is a higher chance of `AbstractHttpRequestHandler` calling `fail` or `failAndReset` multiple times. There is no point in doing so, so added a boolean flag `failed` for safety.
- `HttpResponseWrapper` related changes
  - The original intention of `cancelTimeoutAndLog` was to not log if the response is unexpected. Modified so that if the response is cancelled or the context is cancelled, no logging is done
  - There is probably no reason to not call `close` when a timeout occurs. Unified the fragmented logic of closing the `HttpResponseWrapper`.
- Found that `DefaultRequestLog` is completing futures with the context set on the event loop. We generally don't do this since we strictly validate against setting incorrect `RequestContext`s, so modified to use `withoutContext`.

Result:

- Users can call `ClientRequestContext#cancel` to cancel the ongoing request easily.
- https://github.com/line/armeria/pull/5793 can be prepared for

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 1]
```diff
@@ -428,7 +428,8 @@ private void updateFlags(int flags) {
     private static void completeSatisfiedFutures(RequestLogFuture[] satisfiedFutures, RequestLog log,
                                                  RequestContext ctx) {
         if (!ctx.eventLoop().inEventLoop()) {
-            ctx.eventLoop().execute(() -> completeSatisfiedFutures(satisfiedFutures, log, ctx));
+            ctx.eventLoop().withoutContext().execute(
+                    () -> completeSatisfiedFutures(satisfiedFutures, log, ctx));
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: Q) Why was the context removed? Shouldn't the request context be propagated to the callbacks of `satisfiedFutures`?

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java [hunk 2]
```diff
@@ -76,295 +58,233 @@ final class DefaultCancellationScheduler implements CancellationScheduler {
...
-        pendingTimeoutNanos = timeoutNanos;
         this.server = server;
+        this.ticker = ticker;
     }
 
     /**
      * Initializes this {@link DefaultCancellationScheduler}.
      */
     @Override
     public void initAndStart(EventExecutor eventLoop, CancellationTask task) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:97]: Q) Would it be possible allocating a random `EventLoop` in the constructor? If we decided to use `lock`, I don't see we should use the same event loop of a request for `CancellationScheduler`.

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java [hunk 3]
```diff
@@ -76,295 +58,233 @@ final class DefaultCancellationScheduler implements CancellationScheduler {
...
+                return;
+            }
+            state = State.PENDING;
+            startTimeNanos = ticker.read();
+            if (timeoutMode == TimeoutMode.SET_FROM_NOW) {
+                final long elapsedTimeNanos = startTimeNanos - setFromNowStartNanos;
+                timeoutNanos = LongMath.saturatedSubtract(timeoutNanos, elapsedTimeNanos);
+            }
+            if (timeoutNanos != Long.MAX_VALUE) {
+                scheduledFuture = eventLoop().schedule(() -> invokeTask(null), timeoutNanos, NANOSECONDS);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:133]: Nice clean up!

File: core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java [hunk 4]
```diff
@@ -115,14 +121,31 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Ex
...
         return res;
     }
 
+    private static void updateCancellationTask(ClientRequestContext ctx, HttpRequest req,
+                                               DecodedHttpResponse res) {
+        final ClientRequestContextExtension ctxExt = ctx.as(ClientRequestContextExtension.class);
+        if (ctxExt == null) {
+            return;
+        }
+        ctxExt.responseCancellationScheduler().updateTask(cause -> {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java:138]: I suggested removing `updateTask()` method, but that would result in too many changes to the existing code. As long as the implemetation is correct, I'm okay to maintain the current style and focus on implementing the response timeout mode.

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java (hunk 2 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:97]: I imagine usually the event loop will time out within the `HttpResponseWrapper` or `AbstractHttpRequestHandler`. If the event loop for the `CancellationScheduler` is different from the channel event loop, I think there can be unnecessary event loop rescheduling.

Is there a reason you prefer assigning a random event loop?

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: It seemed like we were making an effort to clear the request context when completing logs.

e.g.
https://github.com/line/armeria/blob/72ebfe12abf7804723a490fe7acfe3a45d46089f/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L242-L247

I saw `ContextCancellationTest` failing occasionally due to this, so I modified to use `withoutContext`.

File: core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java (hunk 4 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java:138]: For the record, `updateTask()` doesn't necessarily have to be removed - it can still be used internally since the old `CancellationScheduler#start` acted similarly to `updateTask()`.

It is probably more important to unify and simplify the request/response abortion mechanism. This seems difficult to do at the moment for client-side. In particular, when one attempts this one should take note of 1) consistency between `[req|res].cause` and `log.[req|res]Cause`, 2) setting the correct exception type (e.g. `UnprocessedRequestException`).

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: It was an inevitable choice Netty can handle pending messages in batch mode that can cause a request context leak.

If an action does not perform sequentially in the callbacks of Netty operations, I think a context-aware executor is generally preferred. 

I am not strong here. If you have any other workarounds, it is okay to keep it.

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java (hunk 2 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:97]: The thing is, `ClientRequestContext` relies on `CancellationScheduler` for certain APIs (e.g. `ClientRequestContext#responseTimeoutMillis`). `ClientRequestContext` is assigned an `EventExecutor` after initialization, so `CancellationScheduler` is following this pattern by assigning an `EventExecutor` later on.

Perhaps if the relationship between `EventExecutor` and `ClientRequestContext` is modified, `CancellationScheduler` can be simpler. Obviously, this is out of this PR's scope though

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: > Netty can handle pending messages in batch mode that can cause a request context leak

I see, I think I misunderstood the intention then. My original guess was that if request logs for multiple requests are waited on a single condition, users would easily see a request context leak.

I'm not sure what you mean by batch mode though. Channels are tied to a single thread, and within a single thread as long as the `try...close` syntax is used context leaks weren't possible.

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: > try...close syntax is used context leaks weren't possible.

Context leaks could happen even with `try...close`. 
`if request logs for multiple requests are waited on a single condition` seems to mean the same thing, but I'll explain it just in case.

Say we have a simple service that adds a callback to `response.whenComplete()`.
```java
try (ctx.push()) {
   val response = HttpResponse.streaming();
   response.write(...);
   response.whenComplete().thenAccept(unused -> {
      // Do something else.
   })
}
```

- The code pushes `ctx` temporary using `try with resource` block.
- `response.whenComplete()` may be completed when `WriteFutureListener.operationComplete()` is invoked if it is a unary type.
https://github.com/line/armeria/blob/cd1973b35fc0326bf7369e1a351d74a1434b911f/core/src/main/java/com/linecorp/armeria/server/AggregatedHttpResponseHandler.java#L92
- Netty `Channel` stores write requests to `ChannelOutboundBuffer`.
- `response.write()` for `request_1` to `request_9` are added to `ChannelOutboundBuffer` and.
- When `response.write()` for `request_10` is written, the channel can flush the outbound data so `request_1` to `request_10` are flushed in the same call (task).
- Without `try (SafeCloseable ignored = RequestContextUtil.pop())`, the callbacks of `request_1-9.whenComplete()` will see the context of `request_10`.

It would be impossible to leak the context if `completeSatisfiedFutures()` is executed asynchronously. So I thought it was safe to remove `.withoutContext()`. 

However, as you said, contexts are not propagated in some cases. I agree with keeping the current style in terms of consistency.

File: core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java [hunk 5]
```diff
@@ -89,18 +111,18 @@ public void run(Throwable cause) { /* no-op */ }
...
+     */
+    void updateTask(CancellationTask cancellationTask);
 
     enum State {
         INIT,
-        INACTIVE,
-        SCHEDULED,
-        FINISHING,
-        FINISHED
+        PENDING,
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java:124]: How about `SCHEDULED`? `PENDING` sounds a like a status between `init()` and `start()`.

File: core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java [hunk 6]
```diff
@@ -110,7 +132,9 @@ interface CancellationTask {
         /**
          * Returns {@code true} if the cancellation task can be scheduled.
          */
-        boolean canSchedule();
+        default boolean canSchedule() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java:135]: nit) If so, should we annotate `@FunctionalInterface`?

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java (hunk 2 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:97]: It was just a random idea but I am satisfied with the current implementation.

**Review by ikhoon — COMMENTED** [review]
Overall, looks nice. Left minor comments and a corner case.

File: core/src/test/java/com/linecorp/armeria/common/ContextPushHookTest.java [hunk 7]
```diff
@@ -135,7 +135,9 @@ void shouldRunHooksWhenContextIsPushed() {
             hookEvents.clear();
             response = client.get("http://foo.com:" + server.httpPort() + "/virtualhost");
             assertThat(response.status()).isEqualTo(HttpStatus.OK);
-            assertThat(hookEvents).containsExactly(
+            // we don't do containsExactly here because there is no easy way to guarantee that
+            // all context hooks from the previous request have been completed
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply core/src/test/java/com/linecorp/armeria/common/ContextPushHookTest.java:139]: Let me just revert this change and handle separately if needed since this is a change independent from this PR

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java [hunk 8]
```diff
@@ -76,295 +58,233 @@ final class DefaultCancellationScheduler implements CancellationScheduler {
...
+        lock.lock();
+        try {
+            if (state != State.INIT) {
+                return;
+            }
+            state = State.PENDING;
+            startTimeNanos = ticker.read();
+            if (timeoutMode == TimeoutMode.SET_FROM_NOW) {
+                final long elapsedTimeNanos = startTimeNanos - setFromNowStartNanos;
+                timeoutNanos = LongMath.saturatedSubtract(timeoutNanos, elapsedTimeNanos);
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:130]: Negative numbers are already handled by netty:

https://github.com/netty/netty/blob/59ccbd9fd692b11f858bfbfe0ea79de364792a17/common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java#L201

Added a `Long#max` call anyways though

File: core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java (hunk 5 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java]: Fixed, note that technically calling `start()` doesn't guarantee a task has been scheduled if the timeout is `Long.MAX_VALUE`

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: Thanks for the explanation 👍 

> When response.write() for request_10 is written, Channel can flush the outbound data so request_1 to request_10 are flushed in the same call (task).

I understood that even if `request_1` to `request_10` is flushed together in a single task, they will be executed sequentially.
e.g.
```java
ClientRequestContext ctx1 = null;
ClientRequestContext ctx2 = null;
// represents a single task
CommonPools.workerGroup().execute(() -> {
    // request 1
    try (SafeCloseable sf = ctx1.push()) {
        ctx1.logBuilder().endResponse();   
    }
    // request 2
    try (SafeCloseable sf = ctx2.push()) {
        ctx2.logBuilder().endResponse();
    }
});
```

>  I agree with keeping the current style in terms of consistency.

👍  for me as well. Ultimately, my thought was that it depends more on how users are using the API - for this reason it is probably more important that the context is pushed (or not) consistently.

e.g. I can imagine the following scenario
```java
HttpResponseWriter res1 = null;
HttpResponseWriter res2 = null;
CompletableFuture.allOf(ctx1.log().whenComplete(), ctx2.log().whenComplete())
                 .handle((v, t) -> {
                     res1.close();
                     res2.close();
                     // when the res is completed, the same thread may call ctx.push somewhere
                     return null;
                 });
```

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java]: Thanks for fixing it. `timeoutNanos` is also exposed to the userland via `ctx.responseTimeoutMillis()`. 0 would be better than a negative number.  
https://github.com/line/armeria/blob/2367de299f71478bd93a618bd40b70a3676656c0/core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java#L824-L826

**Comment by ikhoon** [comment]
https://github.com/line/armeria/pull/5156 has been merged. Would you mind resolving the conflicts?

**Review by ikhoon — APPROVED** [review]
Nice work, @jrhee17! 👍💯

File: core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java [hunk 9]
```diff
@@ -26,10 +26,16 @@
...
+        }
         return new DefaultCancellationScheduler(timeoutNanos, false);
     }
 
     static CancellationScheduler ofServer(long timeoutNanos) {
+        if (timeoutNanos == 0) {
+            timeoutNanos = Long.MAX_VALUE;
+        }
         return new DefaultCancellationScheduler(timeoutNanos, true);
     }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java:40]: What should we do when a user specified `Long.MAX_VALUE`? Should we use `Long.MAX_VALUE - 1` if we handle `Long.MAX_VALUE` specially? :thinking:

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java]: Yeah, but correctness matters :rofl:

**Review by trustin — APPROVED** [review]
Nice clean-up!

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java [hunk 10]
```diff
@@ -76,299 +58,233 @@ final class DefaultCancellationScheduler implements CancellationScheduler {
...
+    @Nullable
+    private volatile Throwable cause;
+    private final Ticker ticker;
+    private final ReentrantShortLock lock = new ReentrantShortLock();
+
+    private final boolean server;
     @Nullable
     private volatile CancellationFuture whenCancelling;
     @Nullable
     private volatile CancellationFuture whenCancelled;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:75]: Because `whenCancelled()`, which creates this future, is called almost immediatly after this instance is created, how about creating this future directory at construction? 
`private final CancellationFuture whenCancelled = new CancellationFuture();`

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java [hunk 11]
```diff
@@ -379,141 +295,72 @@ public Throwable cause() {
...
-                    break;
-                }
-                final Runnable newPendingTask = () -> {
-                    oldPendingTask.run();
-                    pendingTask.run();
-                };
-                if (pendingTaskUpdater.compareAndSet(this, oldPendingTask, newPendingTask)) {
-                    break;
-                }
+        whenCancelled().thenAccept(cause -> {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:328]: This callback is called for all updating tasks.
How about adding this callback only once and use `this.task` instead of `task` in the callback?

**Review by minwoox — COMMENTED** [review]
Sorry for the late review. Left a couple of suggestions.

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:432]: I didn't expect this change to be so controversial 😅  Let me just revert this change since this isn't really critical to this PR

File: core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java (hunk 11 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:328]: I didn't really understand your comment here.

> This callback is called for all updating tasks.

Technically, this code path will be called when `updateTask` is invoked after the `CancellationScheduler` is in the `FINISHED` state. The initial timed out task won't be invoked here

>  How about adding this callback only once and use this.task instead of task in the callback?

Are you suggesting that we modify `this.task`? My concern is this will create a race condition with `invokeTask()`. Do you mind adding a suggestion block so I can understand your intention?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java:317]: Oops, my bad. I didn't notice the `return` statement in the above `if` condition so I thought `thenAccept` will be attached for all `task`s. 😓

**Review by minwoox — APPROVED** [review]
Thanks! 👍 👍 👍

**Comment by ikhoon** [comment]
@jrhee17 Would you check the test failures?
```
ContextCancellationTest > cancel_beforeDelegate(TestInfo) FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [[creqId=147071ed][http://127.0.0.1:34223/#GET]]
        at com.linecorp.armeria.client.ContextCancellationTest.validateCallbackChecks(ContextCancellationTest.java:289)
        at com.linecorp.armeria.client.ContextCancellationTest.cancel_beforeDelegate(ContextCancellationTest.java:115)

ContextCancellationTest > cancel_beforeConnection(TestInfo) FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [[creqId=456bf5ae][http://foo.com:34223/#POST]]
        at com.linecorp.armeria.client.ContextCancellationTest.validateCallbackChecks(ContextCancellationTest.java:289)
        at com.linecorp.armeria.client.ContextCancellationTest.cancel_beforeConnection(ContextCancellationTest.java:153)
```

**Comment by jrhee17** [comment]
Let me go ahead and merge this PR since it's reliable passing the CI again, thanks for the review all 👍
